File: RubyCivitas/lib/civitas_juego.rb
require_relative "gestor_estados"
require_relative "jugador"
require_relative "mazo_sorpresas"
require_relative "sorpresa"
require_relative "tablero"
require_relative "titulo_propiedad"
require_relative 'jugador_especulador'
require_relative 'sorpresa_ircarcel.rb'
require_relative 'sorpresa_jugadoriracasilla.rb'
require_relative 'sorpresa_pagarcobrar.rb'
require_relative 'sorpresa_porcasahotel.rb'
require_relative 'sorpresa_porjugador.rb'
require_relative 'sorpresa_salircarcel.rb'
require_relative 'sorpresa_jugadorespeculador.rb'
require_relative 'casilla_calle.rb'
require_relative 'casilla_sorpresa.rb'
require_relative 'casilla_impuesto.rb'
require_relative 'casilla_juez.rb'
require_relative 'casilla.rb'
require_relative 'jugador_especulador'
require_relative 'jugador'
require "./dado"
module Civitas
  class CivitasJuego

    def initialize(nombre_jugadores)
      
      @jugadores = Array.new
      for i in 0...nombre_jugadores.size
        nuevo = Jugador.new(nombre_jugadores[i])
        jugadorespeculador = Jugador_especulador.copia(nuevo, 1500)
        @jugadores.push(jugadorespeculador)
        
      end
      
      nuevo = Jugador.new("Daniel")
      @jugadores.push(nuevo)

      # Modificación: todos son especuladores menos 1
      
      @gestor_estados = Gestor_estados.new
      @estado = @gestor_estados.estado_inicial
      @indice_jugador_actual = Dado.instance.quien_empieza(@jugadores.length)
      @mazo = MazoSorpresas.new
      @tablero = Tablero.new
      inicializa_mazo_sorpresas(@tablero)
      inicializa_tablero(@mazo)

    end

    def inicializa_tablero(mazo_s)

      @tablero = Tablero.new(9);
      
      # CASILLA 1
      propiedad_n1 = TituloPropiedad.new("Reyes Catolicos",25,50,1000,2000,500)
      casilla_n1 = Casilla_calle.new(propiedad_n1)
      @tablero.aniade_casilla(casilla_n1)
      
      # CASILLA 2
      propiedad_n2 = TituloPropiedad.new("Callao", 25, 50, 1000, 2000, 500)
      casilla_n2 = Casilla_calle.new(propiedad_n2)
      @tablero.aniade_casilla(casilla_n2)
      
      # SORPRESA 1
      sorpresa_n1 = Casilla_sorpresa.new(mazo_s, "Casilla sorpresa n1")
      @tablero.aniade_casilla(sorpresa_n1)
      
      # CASILLA 3
      propiedad_n3 = TituloPropiedad.new("Pedro Antonio",25,50,1000,2000,500)
      casilla_n3 = Casilla_calle.new(propiedad_n3)
      @tablero.aniade_casilla(casilla_n3)
      
      # CASILLA 4
      propiedad_n4 = TituloPropiedad.new("Catedral",50.0,100,2000,4000,1000)
      casilla_n4 = Casilla_calle.new(propiedad_n4)
      @tablero.aniade_casilla(casilla_n4)
      
      # CASILLA 5
      propiedad_n5 = TituloPropiedad.new("Camino Ronda",50.0,100,2000,4000,1000)
      casilla_n5 = Casilla_calle.new(propiedad_n5)
      @tablero.aniade_casilla(casilla_n5)
      
      # CASILLA 6
      propiedad_n6 = TituloPropiedad.new("Malagueta",50.0,100,2000,4000,1000)
      casilla_n6 = Casilla_calle.new(propiedad_n6)
      @tablero.aniade_casilla(casilla_n6)
      
      # CASILLA 7
      propiedad_n7 = TituloPropiedad.new("Alicante",50.0,100,2000,4000,1000)
      casilla_n7 = Casilla_calle.new(propiedad_n7)
      @tablero.aniade_casilla(casilla_n7)
      
      # CASILLA 8
      propiedad_n8 = TituloPropiedad.new("Xeraco",50.0,100,2000,4000,1000)
      casilla_n8 = Casilla_calle.new(propiedad_n8)
      @tablero.aniade_casilla(casilla_n8)
      
      # CASILLA 9
      propiedad_n9 = TituloPropiedad.new("Albaicin",50.0,100,2000,4000,1000)
      casilla_n9 = Casilla_calle.new(propiedad_n9)
      @tablero.aniade_casilla(casilla_n9)
      
      # CASILLA 10
      propiedad_n10 = TituloPropiedad.new("La Zubia",50.0,100,2000,4000,1000)
      casilla_n10 = Casilla_calle.new(propiedad_n10)
      @tablero.aniade_casilla(casilla_n10)
      
      # CASILLA 11
      propiedad_n11 = TituloPropiedad.new("Zaidin",50.0,100,2000,4000,1000)
      casilla_n11 = Casilla_calle.new(propiedad_n11)
      @tablero.aniade_casilla(casilla_n11)
      
      # CASILLA 12
      propiedad_n12 = TituloPropiedad.new("Palacio Deportes",50.0,100,2000,4000,1000)
      casilla_n12 = Casilla_calle.new(propiedad_n12)
      @tablero.aniade_casilla(casilla_n12)
      
      # IMPUESTO 1
      hacienda = Casilla_impuesto.new("Hacienda!",1000)
      @tablero.aniade_casilla(hacienda)
      
      # DESCANSO 1
      la_nada = Casilla.new("Nada")
      @tablero.aniade_casilla(la_nada)
      
      # SORPRESA 2
      sorpresan2 = Casilla_sorpresa.new(mazo_s,"Casilla sorpresa n2")
      @tablero.aniade_casilla(sorpresan2)
      
      # SORPESA 3
      sorpresan3 = Casilla_sorpresa.new(mazo_s,"Casilla sorpresa n3")
      @tablero.aniade_casilla(sorpresan3)

      @tablero.aniade_juez

    end

    def inicializa_mazo_sorpresas(tablero)
      
      especulador = Sorpresa_jugadorespeculador.new(1500,"[SORPRESA] Conversion ESPECULADOR")
      @mazo.al_mazo(especulador)

      ir_a_carcel = Sorpresa_ircarcel.new(tablero);
      @mazo.al_mazo(ir_a_carcel);

      ir_a_casilla = Sorpresa_jugadoriracasilla.new(tablero,3,"[SORPRESA] Vete a la casilla 3");
      @mazo.al_mazo(ir_a_casilla);

      ir_a_casilla_n2 = Sorpresa_jugadoriracasilla.new(tablero,5,"[SORPRESA] Vete a la casilla 5");
      @mazo.al_mazo(ir_a_casilla_n2);

      salir_carcel = Sorpresa_salircarcel.new(@mazo);
      @mazo.al_mazo(salir_carcel);

      pagar = Sorpresa_pagarcobrar.new(-50,"[SORPRESA] Tienes que pagar 50")
      @mazo.al_mazo(pagar);

      cobrar = Sorpresa_pagarcobrar.new(50,"[SORPRESA] Tienes que cobrar 50");
      @mazo.al_mazo(cobrar);

      por_casa_hotel = Sorpresa_porcasahotel.new(20,"[SORPRESA] Pagas 100 por cada casa y hotel");
      @mazo.al_mazo(por_casa_hotel);

      por_jugador_n1 = Sorpresa_porjugador.new(30,"[SORPRESA] Cada jugador te paga 30");
      @mazo.al_mazo(por_jugador_n1);

      por_jugador_n2 = Sorpresa_porjugador.new(-30,"[SORPRESA] Se te quitan 30 para pagar a un jugador");
      @mazo.al_mazo(por_jugador_n2);

    end

    def contabilizar_pasos_por_salida(jugador_actual)

      while(@tablero.get_por_salida>0)

        jugador_actual.pasa_por_salida

      end
    end

    def pasar_turno

      @indice_jugador_actual = (@indice_jugador_actual + 1 ) % @jugadores.size

    end

    def siguiente_paso_completo(operacion)

      @estado = @gestor_estados.siguiente_estado(@jugadores.at(@indice_jugador_actual), @estado, operacion);

    end

    def construir_casa(ip)

      @jugadores.at(@indice_jugador_actual).construir_casa(ip);

    end

    def construir_hotel(ip)

      @jugadores.at(@indice_jugador_actual).construir_hotel(ip)

    end

    def vender(ip)

      @jugadores.at(@indice_jugador_actual).vender(ip);

    end

    def hipotecar(ip)

      @jugadores.at(@indice_jugador_actual).hipotecar(ip);

    end

    def cancelar_hipoteca(ip)

      @jugadores.at(@indice_jugador_actual).cancelar_hipoteca(ip);

    end

    def salir_carcel_pagando

      @jugadores.at(@indice_jugador_actual).salir_carcel_pagando

    end

    def salir_carcel_tirando

      @jugadores.at(@indice_jugador_actual).salir_carcel_tirando

    end

    def final_del_juego

      for jugador in @jugadores

        if jugador.saldo == 0

          return true

        end
      end

      return false

    end

    def ranking

      ranking = Array.new {Jugador}

      @jugadores.each { |jugadores|

        ranking.push(jugadores)

      }

      ranking.sort.reverse!

    end

    def avanza_jugador

      # jugadorActual y posicion_actual
      jugador_actual = @jugadores.at(@indice_jugador_actual)
      posicion_actual = jugador_actual.num_casilla_actual

      # tirada y posicionNueva

      tirada = Dado.instance.tirar
      posicion_nueva = @tablero.nueva_posicion(posicion_actual, tirada)

      # casilla
      casilla = @tablero.get_casilla(posicion_nueva)

      # NOTA
      contabilizar_pasos_por_salida(jugador_actual)

      jugador_actual.mover_a_casilla(posicion_nueva)
      casilla.recibe_jugador(@indice_jugador_actual, @jugadores)

      # Por si acaso, pasa por salida
      contabilizar_pasos_por_salida(jugador_actual)

    end

    def get_jugador_actual

      @jugadores[@indice_jugador_actual]

    end

    def get_casilla_actual

      @tablero.get_casilla(@jugadores[@indice_jugador_actual].num_casilla_actual)

    end

    def siguiente_paso

      jugador_actual = @jugadores.at(@indice_jugador_actual)
      
      operacion = @gestor_estados.operaciones_permitidas(jugador_actual, @estado)

      if(operacion == Operaciones_juego::PASAR_TURNO)

        pasar_turno
        siguiente_paso_completo(operacion)
      elsif (operacion == Operaciones_juego::AVANZAR)

        avanza_jugador
        siguiente_paso_completo(operacion)

      end

      operacion

    end

    def comprar

      jugador_actual = get_jugador_actual
      casilla = get_casilla_actual
      titulo = casilla.titulo_propiedad

      res = jugador_actual.comprar(titulo)

    end

  end
end

File: civitasjava/src/Civitas/CivitasJuego.java
package civitas;

import java.util.ArrayList;
import civitas.TituloPropiedad;
import java.util.Collections;

public class CivitasJuego {
    private int indiceJugadorActual;
    private EstadosJuego estado;
    private GestorEstados gestor;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private ArrayList<Jugador> jugadores;
    private OperacionesJuego operacion;
    
    /**
     * Cambia de posicion al jugador
     */
    private void avanzaJugador(){
        Jugador jugadorActual = getJugadorActual();
        int posicionActual = jugadorActual.getNumCasillaActual();
        int tirada = Dado.getInstance().tirar();
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        Casilla casilla = tablero.getCasilla(posicionNueva);
        this.contabilizarPasosPorSalida(jugadorActual);
        jugadorActual.moverACasilla(posicionNueva);
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        this.contabilizarPasosPorSalida(jugadorActual);
    };
	
    /**
     * Cancela la hipoteca de un jugador
     * @param ip : propiedad de la que se quiere cancelar la hipoteca
     * @return 
     */
    public boolean cancelarHipoteca(int ip){
        return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
};

    /**
     * Constructor de la clase que:
     * 1) Inicializa el atributo jugadores creando y añadiendo un jugador por 
     *    cada nombre suministrado como parametro
     * 2) Crear el gestor de estados
     * 3) Inicializa el indice del jugador actual
     * 4) Crea el mazo de sorpresas
     * 5) Inicializa el tablero
     * 6) Inicializa el mazo de sorpresas
     * @param nombres : nombre de los jugadores que se quieren añadir
     */
    public CivitasJuego(ArrayList<String> nombres){
        jugadores = new ArrayList<>();
        for(int i=0; i<nombres.size(); i++){
            jugadores.add(new Jugador(nombres.get(i)));
        }
       
        gestor = new GestorEstados();
        gestor.estadoInicial();
		
        indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());

        estado = EstadosJuego.INICIO_TURNO;

        mazo = new MazoSorpresas(true);
        
        mazo = new MazoSorpresas(true);
        inicializarTablero(mazo);
        inicializarMazoSorpresas(tablero);
        
		
    };
    
    /**
     * Compra la propiedad de la casilla en la que se encuentra
     * @return res
     */
    public boolean comprar(){
        boolean res = false;
        
        Jugador jugadoractual = jugadores.get(indiceJugadorActual);
        int numCasillaActual = jugadoractual.getNumCasillaActual();
        Casilla casilla = tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo = ((CasillaCalle)casilla).getTituloPropiedad();
        res = jugadoractual.comprar(titulo);
        
        return res;
    };
	

    /**
     * Delega en el metodo construirCasa(ip) de Jugador
     * @param ip : propiedad donde se quiere construir la casa
     * @return 
     */
    public boolean construirCasa(int ip){
        return jugadores.get(indiceJugadorActual).construirCasa(ip);
    }
	
    /**
     * Delega en el metodo construirHotel de Jugador
     * @param ip : propiedad donde se quiere construir el hotel
     * @return 
     */
    public boolean construirHotel(int ip){
        return jugadores.get(indiceJugadorActual).construirHotel(ip);
    };

    /**
     * Mientras el método getPorSalida del tablero devuelva un valor mayor que 
     * 0 se llama al método pasaPorSalida del jugador pasado como parámetro. 
     * Al llamar a ese método el jugador actual cobra por todas las veces que 
     * ha pasado por la salida en su turno actual.
     * @param jugadorActual : jugador que va a cobrar la salida
     */
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        if(tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        }
    };

    /**
     * Devuelve true si alguno de los jugadores esta en bancarrota
     * @return fin
     */
    public boolean finalDelJuego(){
        boolean fin = false;

        for(int i=0; i<jugadores.size(); i++){
            
            if(jugadores.get(i).enBancarrota()){
                fin = true;

            }
        }

        return fin;
    };
	
    /**
     * Delega en el metodo getCasilla de Tablero
     * @return 
     */
    public Casilla getCasillaActual(){
        return tablero.getCasilla(getJugadorActual().getNumCasillaActual());
    };

    /**
     * Devuelve el jugador actual
     * @return 
     */
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    };

    /**
     * Delega en el metodo hipotecar(ip) de Jugador
     * @param ip : indice de la propiedad
     * @return 
     */
    public boolean hipotecar(int ip){
        return jugadores.get(indiceJugadorActual).hipotecar(ip);
    };

    public String infoJugadorTexto(){
	return (jugadores.get(indiceJugadorActual).toString() + "\n"
	+"Casilla actual: " +getCasillaActual().toString() +"\n--------------");
    };
	
    /**
     * Inicializa el mazo de sorpresas rellenando el mazosorpresas con sorpresas
     * @param tablero : 
     */
    private void inicializarMazoSorpresas(Tablero tablero){
        
      
        mazo.alMazo(new SorpresaPagarCobrar(250, "Cobras 250"));
        
        //mazo.alMazo(new SorpresaIrCasilla(tablero, 0, "Avanzas a la casilla de salida"));
        
        mazo.alMazo(new SorpresaPagarCobrar(-250, "Pagas 250"));
        
        mazo.alMazo(new SorpresaPorJugador(250, "Cobras 250"
                + "por cada jugador"));
        
        mazo.alMazo(new SorpresaSalirCarcel(mazo, "Obtienes la llave"));
        
        mazo.alMazo(new SorpresaPorCasaHotel(250, "Cobras 250"
                + "por cada hotel"));
    };

    /**
     * Se inicializa el tablero completando cada una de sus casillas
     * @param mazo :
     */
    private void inicializarTablero(MazoSorpresas mazo){
        
        tablero = new Tablero(5);
        this.mazo = mazo;
        
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Ogijares",110,
                (float) 1.29,680,1150,240)));

        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Armilla",110,
                (float) 1.29,680,1150,240)));
        
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Otura",110,
                (float) 1.29,680,1150,240)));

        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa1"));

        tablero.añadeCasilla(new CasillaImpuesto((float)1200.0, "Multado"));
        
        tablero.añadeJuez();
        
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("La chana",110,
                (float) 1.29,680,1150,240)));

        tablero.añadeCasilla(new Casilla("Parking"));

        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Albolote", 150,
                (float) 1.45, 860, 1850, 320 )));
        
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Peligros",110,
                (float) 1.29,680,1150,240)));
        
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Pulianas",110,
                (float) 1.29,680,1150,240)));

        tablero.añadeCasilla(new CasillaSorpresa(mazo, "sorpresa 2"));

        tablero.añadeCasilla(new CasillaImpuesto((float)1235.0, "Multilla"));
    };
	
    /**
     * Se pasa el turno de un jugador al siguiente
     * Si es el ultimo se pasa al primero 
     */
    private void pasarTurno(){
        indiceJugadorActual++;
        if(indiceJugadorActual == jugadores.size())
            indiceJugadorActual = 0;
    };

    /**
     * Produce una lista ordenada de jugadores en funcion de su saldo
     * @return lista
     */
    public ArrayList<Jugador> ranking(){
        /*ArrayList<Jugador> ranking = new ArrayList<Jugador>();

        for(int i=0; i<jugadores.size(); i++){
            ranking.add(jugadores.get(i));
        }

        for(int i=0; i<ranking.size(); ++i){
            for(int j=i+1; j<ranking.size(); ++j){
                if(ranking.get(i).compareTo(ranking.get(j)) < 0){
                        Jugador aux = new Jugador(ranking.get(i));
                        ranking.set(i, ranking.get(j));
                        ranking.set(j,aux);
                }
            }
        }
        return ranking;*/
        ArrayList<Jugador> lista=jugadores;
        lista.sort((o1,o2) -> o1.compareTo(o2));
        Collections.reverse(lista);
        return lista;
    };
	
    /**
     * Se muestra el ranking de jugadores
     */
    public void mostrarRanking(){
        for(int i=0; i<ranking().size(); i++){
            System.out.println(i +".-" +ranking().get(i));
        }
    }

    /**
     * Delega em el metodo salircarcelpagando de Jugador
     * @return 
     */
    public boolean salirCarcelPagando(){
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
    };

    /**
     * Delega en el metodo salircarceltirando de Jugador
     * @return 
     */
    public boolean salirCarcelTirando(){
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
    };
    
    /**
     * Muestra cual es el siguiente paso elegido
     * @return operacion
     */
    public OperacionesJuego siguientePaso(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion = gestor.operacionesPermitidas(jugadorActual,
                estado);

        if(operacion == OperacionesJuego.PASAR_TURNO){
            this.pasarTurno();
            this.siguientePasoCompletado(operacion);
        }
        else if(operacion == OperacionesJuego.AVANZAR){
            this.avanzaJugador();
            this.siguientePasoCompletado(operacion);
        }

        return operacion;
    }

    /**
     * Se actualiza el estado del juego obteniendo el siguiente  estado del 
     * gestor de estados (método siguienteEstado).
     * Para ello es necesario obtener el jugador actual.
     * @param operacion 
     */
    public void siguientePasoCompletado(OperacionesJuego operacion){
	estado = gestor.siguienteEstado(jugadores.get(indiceJugadorActual),
                estado, operacion);
    }

    /**
    * Delega en el metodo vender(ip) de Jugador
    * @param ip : indice de la propiedad
    * @return 
    */
    public boolean vender(int ip){
        return jugadores.get(indiceJugadorActual).vender(ip);
    };
 
      
}

